### PR TITLE
Add a YARP device example for using YarpCameraBridge and ArucoDetector in Gazebo simulation

### PR DESCRIPTION
**Pull request by prashanthr05**
P.S. The large file number is mainly due to the addition of gazebo models for Aruco markers.

As a result of this PR, we are able to do this,
|YarpCameraBridge based Aruco detection in simulation|
|:--:|
|Screenshot from 2022-01-28 17-46-42|

### CODE REVIEW

**Comment by prashanthr05** [comment]
Maybe it's better to separate this PR into two, one handling code changes and another handling the Gazebo aruco marker changes?

File: devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+### Aruco marker Gazebo models 
+
+- **Dictionary** DICT_4x4_50 
+- **Side** 80.6 mm
+- **IDs** 0, 2, 3, 6, 7, 8, 9, 12, 13, 15, 16, 18, 20
+
+
+
```
Comment (GiulioRomualdi) [inline devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/README.md:8]:
```suggestion

```

File: devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ To use, update `GAZEBO_MODEL_PATH` in .bashrc
+
+
+
```
Comment (GiulioRomualdi) [inline devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/README.md:12]:
```suggestion

```

File: devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+```
+export GAZEBO_MODEL_PATH=${GAZEBO_MODEL_PATH}:<path-to-KinDynFusion>/gazebo/models
+```
+
+
+
```
Comment (GiulioRomualdi) [inline devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/README.md:18]:
```suggestion

```

File: devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/aruco_marker_4x4_1000-0/materials/scripts/aruco_marker_4x4_1000-0.material [hunk 4]
```diff
@@ -0,0 +1,13 @@
+material ArucoMarker4x4_1000_0/Marker
```
Comment (GiulioRomualdi) [inline devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/aruco_marker_4x4_1000-0/materials/scripts/aruco_marker_4x4_1000-0.material:1]: I've noticed that the number of file in the material folder is really huge. Where did you find them? Are there available somewhere else? If yes we could fetch them from cmake and avoid to store them in this repo

File: devices/examples/YarpCameraBridgeArucoTestDevice/src/YarpCameraBridgeArucoTestDevice.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+{
+    constexpr auto logPrefix = "[YarpCameraBridgeArucoTestDevice::run]";
+
+    const double time = BipedalLocomotion::clock().now().count();
+    double imgRcvTime;
+    bool ok = m_robotCameraBridge->getColorImage("rgbd_cam", m_rgbImg, imgRcvTime);
+    if (ok)
+    {
+        cv::imshow("rgb", m_rgbImg);
+        cv::waitKey(1);
```
Comment (GiulioRomualdi) [inline devices/examples/YarpCameraBridgeArucoTestDevice/src/YarpCameraBridgeArucoTestDevice.cpp:148]: This is just a curiosity, what `cv::waitKey()` does?  and why this is required?

**Review by GiulioRomualdi — COMMENTED** [review]
I've just some concerns regarding the `material` folder. It would be nice to fetch the content from the original source so we avoid to add them to the repo.

**Comment by GiulioRomualdi** [comment]
Regarding the generation of the markers I found this: https://www.pyimagesearch.com/2020/12/14/generating-aruco-markers-with-opencv-and-python/

File: devices/examples/YarpCameraBridgeArucoTestDevice/src/YarpCameraBridgeArucoTestDevice.cpp (hunk 5 above)
Comment (prashanthr05) [reply devices/examples/YarpCameraBridgeArucoTestDevice/src/YarpCameraBridgeArucoTestDevice.cpp:148]: waitKey simply adds a delay in milliseconds and a small delay is necessary to be added for `imshow` to display the image in the corresponding window. For more information, see https://docs.opencv.org/2.4/modules/highgui/doc/user_interface.html#imshow.

File: devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/aruco_marker_4x4_1000-0/materials/scripts/aruco_marker_4x4_1000-0.material (hunk 4 above)
Comment (prashanthr05) [reply devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/aruco_marker_4x4_1000-0/materials/scripts/aruco_marker_4x4_1000-0.material:1]: I had used aruco markers generated using the [ArucoGen online software](https://chev.me/arucogen/) to generate markers from a 4x4 dictionary for the robot experiments. So, for the simulation, I generated the same markers and used them as the material for the gazebo models, taking inspiration from https://github.com/joselusl/aruco_gazebo.

Comment (prashanthr05) [reply devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/aruco_marker_4x4_1000-0/materials/scripts/aruco_marker_4x4_1000-0.material:1]: > If yes we could fetch them from cmake and avoid to store them in this repo

We can do that.

Comment (prashanthr05) [reply devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/aruco_marker_4x4_1000-0/materials/scripts/aruco_marker_4x4_1000-0.material:1]: Where can we host that content? a separate repo? or shared storage?

Comment (prashanthr05) [reply devices/examples/YarpCameraBridgeArucoTestDevice/gazebo/models/aruco_marker_4x4_1000-0/materials/scripts/aruco_marker_4x4_1000-0.material]: I have removed the models in this commit https://github.com/ami-iit/bipedal-locomotion-framework/pull/504/commits/9dc82b8384404c7c0153c8ce75e5e63592012f2f.

If we decide where to host them, I shall update the documentation accordingly on fetching and using them.

**Comment by GiulioRomualdi** [comment]
I'm going to close this PR since no one is working on this. If required in the future we can revive it :)

**Pull request closed without merging**